File: src/class036/C7_PreorderInorderBuildBinaryTree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# 使用先序与中序遍历序列构造二叉树
def buildTree(preorder, inorder):
    if not preorder or not inorder or len(preorder) != len(inorder):
        return None
    
    # 创建哈希映射以优化查找速度
    idx_map = {value: idx for idx, value in enumerate(inorder)}
    
    # 辅助函数，通过递归构造子树
    def construct(left, right):
        if left > right:
            return None
        
        # 根节点为先序遍历的第一个元素
        root_value = preorder.pop(0)
        root = TreeNode(root_value)
        
        # 根据根节点在中序遍历中的位置，分割左右子树
        index = idx_map[root_value]
        
        # 构造左子树和右子树
        root.left = construct(left, index - 1)
        root.right = construct(index + 1, right)
        
        return root
    
    return construct(0, len(inorder) - 1)

File: src/class036/C7_PreorderInorderBuildBinaryTree.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

// 使用先序与中序遍历序列构造二叉树
// 测试链接: https://leetcode.cn/problems/construct-binary-tree-from-preorder-and-inorder-traversal/

// 不提交这个类
class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

// 提交如下的方法
TreeNode* buildTree(vector<int>& pre, vector<int>& in) {
    if (pre.empty() || in.empty() || pre.size() != in.size()) {
        return nullptr;
    }
    unordered_map<int, int> map;
    for (int i = 0; i < in.size(); i++) {
        map[in[i]] = i;
    }
    return construct(pre, 0, pre.size() - 1, in, 0, in.size() - 1, map);
}

TreeNode* construct(const vector<int>& pre, int l1, int r1, const vector<int>& in, int l2, int r2, unordered_map<int, int>& map) {
    if (l1 > r1) {
        return nullptr;
    }
    TreeNode* head = new TreeNode(pre[l1]);
    if (l1 == r1) {
        return head;
    }
    int k = map[pre[l1]];
    head->left = construct(pre, l1 + 1, l1 + k - l2, in, l2, k - 1, map);
    head->right = construct(pre, l1 + k - l2 + 1, r1, in, k + 1, r2, map);
    return head;
}
